### ISSUE TITLE
twisted.internet.test.test_process.ProcessTestsBuilder_AsyncioSelectorReactorTests.test_childConnectionLost fails on python 3 on OS X

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8754]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#8754|
|Type|defect|
|Created|2016-08-13 07:13:53Z|

```
[ERROR]
Traceback (most recent call last):
  File "/Users/glyph/Projects/Twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/test/test_process.py", line 650, in lostSecond
    transport.write(b'1\n')
  File "/Users/glyph/Projects/Twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/process.py", line 903, in write
    self.pipes[0].write(data)
  File "/Users/glyph/Projects/Twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/process.py", line 197, in write
    abstract.FileDescriptor.write(self, data)
  File "/Users/glyph/Projects/Twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/abstract.py", line 355, in write
    self.startWriting()
  File "/Users/glyph/Projects/Twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/abstract.py", line 442, in startWriting
    self.reactor.addWriter(self)
  File "/Users/glyph/Projects/Twisted/build/py35-coverage-posix/lib/python3.5/site-packages/twisted/internet/asyncioreactor.py", line 172, in addWriter
    False)
  File "/Homebrew/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/selector_events.py", line 275, in add_writer
    (None, handle))
  File "/Homebrew/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/selectors.py", line 540, in register
    self._kqueue.control([kev], 0, 0)
builtins.BrokenPipeError: [Errno 32] Broken pipe

twisted.internet.test.test_process.ProcessTestsBuilder_AsyncioSelectorReactorTests.test_childConnectionLost
```

<details><summary>Searchable metadata</summary>

```
trac-id__8754 8754
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1471072433138924 1471072433138924
changetime__1471133134006504 1471133134006504
version__None None
owner__GitHub__noreply_____ GitHub <noreply@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

The problem is this (from all extant versions of Python 3):

```python
>>> OSError
<class 'OSError'>
>>> IOError
<class 'OSError'>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

No idea why I'm seeing it so reliably locally but not seeing it on the buildbot...

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph removed owner|
|-|-|

PR at https://github.com/twisted/twisted/pull/493

No new tests because this was causing the tests to fail reliably on my machine, and with this fix they pass reliably, so I believe it ought to be covered by the existing suite.  Not sure why we're not seeing this in CI, though!

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_4"></a>|@markrwilliams commented|
|-|-|

Crazy as it seems, here's a citation into the PEP that started it all:

https://www.python.org/dev/peps/pep-3151/#step-1-coalesce-exception-types

I put a minor comment into the PR, but I think this is ready to merge once the builds are green.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|GitHub set owner to GitHub<br>GitHub set status to `closed`|
|-|-|

In changeset 684d88e7b73cb4af011edc6a6a7d435508c30b99
```
#!CommitTicketReference repository="" revision="684d88e7b73cb4af011edc6a6a7d435508c30b99"
Merge pull request #493 from twisted/8754-kqueue-error-handling

Author: glyph

Reviewer: markrwilliams 

Fixes: #8754

Handle the correct exception type raised by asyncio on OS X when pipes are broken.
```
